core/trino-main/src/main/java/io/trino/execution/DropColumnTask.java

```diff
@@ -87,17 +90,66 @@ public ListenableFuture<Void> execute(
             return immediateVoidFuture();
         }
 
-        if (metadata.getColumnMetadata(session, tableHandle, columnHandle).isHidden()) {
+        List<String> field = statement.getField().getOriginalParts().subList(1, statement.getField().getOriginalParts().size()).stream()
+                .map(Identifier::getValue)
+                .collect(toImmutableList());
+        ColumnMetadata columnMetadata = metadata.getColumnMetadata(session, tableHandle, columnHandle);
+        if (columnMetadata.isHidden()) {
             throw semanticException(NOT_SUPPORTED, statement, "Cannot drop hidden column");
         }
-
-        if (metadata.getTableMetadata(session, tableHandle).getColumns().stream()
-                .filter(info -> !info.isHidden()).count() <= 1) {
-            throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only column in a table");
+        if (field.isEmpty()) {
+            if (metadata.getTableMetadata(session, tableHandle).getColumns().stream()
+                    .filter(info -> !info.isHidden()).count() <= 1) {
+                throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only column in a table");
+            }
+            metadata.dropColumn(session, tableHandle, columnHandle);
+        }
+        else {
+            if (columnMetadata.getType() instanceof RowType rowType) {
+                MatchResult result = findField(rowType.getFields(), field);
+                if (result.matched > 0 && result.fields - result.matched == 0) {
```

Comment:
This can perhaps be written as "a walk"-like iteration

```java
Type containingType = columnMetadata.getType();
            for (String fieldName : fieldPath) {
                if (containingType instanceof RowType rowType) {
                    List<RowType.Field> candidates = rowType.getFields().stream()
                            // case-sensitive match
                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equals(fieldName))
                            .collect(toImmutableList());
                    if (candidates.size() > 1) {
                        throw semanticException(COLUMN_NOT_FOUND, statement, "Field path %s within %s is ambiguous", fieldPath, columnMetadata.getType());
                    }
                    if (candidates.size() == 1) {
                        containingType = getOnlyElement(candidates).getType();
                        continue;
                    }
                }
                // TODO: Support array and map types
                throw semanticException(
                        NOT_SUPPORTED,
                        statement,
                        "Cannot resolve field '%s' within %s type when dropping %s in %s",
                        fieldName,
                        containingType,
                        fieldPath,
                        columnMetadata.getType());
            }

            // At this point containingType is known to be a RowType
            if (((RowType) containingType).getFields().size()==1) {
                throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only field in a row type");
            }

            metadata.dropField(session, tableHandle, columnHandle, fieldPath);
```

Reply:
Did you ignore `isColumnExists` option intentionally? 

> // At this point containingType is known to be a RowType

This isn't ensured in the current logic though.

Reply:
> Did you ignore `c` option intentionally?

good point
not sure if this should get applied to the leaf field only, or to every step on the path.

if to every step, adding something like
```
if (isColumnExists && candidates.size() == 0) {
  return immediate(..)
```

would suffice

Reply:
> This isn't ensured in the current logic though.

right, i made a mistake. perhaps

```java
            RowType containingType = null;
            Type currentType = columnMetadata.getType();
            for (String fieldName : fieldPath) {
                if (currentType instanceof RowType rowType) {
                    List<RowType.Field> candidates = rowType.getFields().stream()
                            // case-sensitive match
                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equals(fieldName))
                            .collect(toImmutableList());
                    if (candidates.size() > 1) {
                        throw semanticException(COLUMN_NOT_FOUND, statement, "Field path %s within %s is ambiguous", fieldPath, columnMetadata.getType());
                    }
                    if (candidates.size() == 1) {
                        RowType.Field rowField = getOnlyElement(candidates);
                        containingType = rowType;
                        currentType = rowField.getType();
                        continue;
                    }
                    if (statement.isColumnExists() && candidates.size() == 0) {
                        // TODO should we allow only the leaf not to exist, or any path component?
                        return immediateVoidFuture();
                    }
                }
                // TODO: Support array and map types
                throw semanticException(
                        NOT_SUPPORTED,
                        statement,
                        "Cannot resolve field '%s' within %s type when dropping %s in %s",
                        fieldName,
                        currentType,
                        fieldPath,
                        columnMetadata.getType());
            }

            requireNonNull(containingType, "containingType is null");
            if (containingType.getFields().size() == 1) {
                throw semanticException(NOT_SUPPORTED, statement, "Cannot drop the only field in a row type");
            }
```

Reply:
Thanks for preparing the code! Updated.